scripts/release/prod/update_releases_page

```diff
@@ -0,0 +1,176 @@
+    text = ""
+    for x in rset:
+        path = x["Key"]
+        pre, fname = path.rsplit("/", 1)
+        if fname.endswith(".asc"):
+            continue
+        if fname.endswith(".sig"):
+            continue
+        if fname.startswith("hashes"):
+            ob = s3.get_object(Bucket=staging_bucket, Key=path)
+            text += ob["Body"].read().decode()
+    return text
+
+def read_hashes(fin):
+    by_fname = {}
+    for line in fin:
+        if not line:
+            continue
+        line = line.strip()
+        if not line:
+            continue
+        if line[0] == "#":
+            continue
+        hashstr, fname = line.split()
+        ob = by_fname.get(fname)
+        if not ob:
+            ob = {}
+            by_fname[fname] = ob
+        if len(hashstr) == 32:
+            ob["md5"] = hashstr
+        elif len(hashstr) == 64:
+            ob["sha256"] = hashstr
+        elif len(hashstr) == 128:
+            ob["sha512"] = hashstr
+    return by_fname
+
+def objects_by_fname(they):
+    out = {}
+    for x in they:
+        path = x["Key"]
+        if path.endswith("/"):
+            continue
+        parts = path.rsplit("/", 1)
+        fname = parts[-1]
+        out[fname] = x
+    return out
+
+def main(channel):
+    s3 = boto3.client("s3")
+    staging_response = s3.list_objects_v2(Bucket=staging_bucket, Prefix="releases/" + channel + "/", MaxKeys=100)
+    rset = get_stage_release_set(staging_response)
+    hashftext = get_hashes_data(s3, rset)
+    fhashes = read_hashes(hashftext.splitlines())
+    files = release_set_files(rset)
+    releases_response = s3.list_objects_v2(Bucket=releases_bucket)
+    release_files = objects_by_fname(releases_response["Contents"])
+
+    def get_furl(fname, skey):
+        rfpath = release_files.get(fname)
+        if rfpath is not None:
+            return releases_prefix + rfpath["Key"]
+        else:
+            return staging_prefix + skey
+
+    tbody = ["<tbody>\n"]
+    for fname, info in files.items():
+        furl = get_furl(fname, info['file'])
+        ftext = '<div class="fname"><a href="{}">{}</a></div>'.format(furl, fname)
+        sig = info.get(".sig")
+        stext = ""
+        if sig is not None:
+            sfname = sig["Key"].rsplit("/", 1)[-1]
+            surl = get_furl(sfname, sig["Key"])
+            stext = '<a href="{}">.sig</a>'.format(surl)
+        size = info.get("Size", "")
+        hashes = fhashes.get(fname)
+        if hashes:
+            for hn in ("md5", "sha256", "sha512"):
+                hv = hashes.get(hn)
+                if hv:
+                    ftext += '<div class="hash {}">{}</div>\n'.format(hn, hv)
+        if not hashes and not stext:
+            continue
+        tbody.append("<tr><td>{}</td><td>{}</td><td>{}</td></tr>\n".format(ftext, size, stext))
+    tbody.append("</tbody>\n\n")
+    tbody.append('<tbody><tr class="spacer"><td></td></tr></tbody>\n\n')
+
+    match_string = "<!--" + channel + "-->"
```

Comment:
this will probably work, but it could break in a bad way. If this were automated a runaway loop could leave a page full of garbage. There's nothing to prevent from adding the same record many times. I think a safer strategy would be to accumulate data to a separate structure that parses well (CSV file, JSON file, DynamoDB, ...) and then parse that and emit the whole index.html based on the source data.

Reply:
Will change.